File: SourceAFIS/Extraction/Filters/LocalHistogram.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading.Tasks;
using SourceAFIS.General;

namespace SourceAFIS.Extraction.Filters
{
    public sealed class LocalHistogram
    {
        public short[, ,] Analyze(BlockMap blocks, byte[,] image)
        {
            short[, ,] histogram = new short[blocks.BlockCount.Height, blocks.BlockCount.Width, 256];
            foreach (var block in blocks.AllBlocks)
            {
                RectangleC area = blocks.BlockAreas[block];
                for (int y = area.Bottom; y < area.Top; ++y)
                    for (int x = area.Left; x < area.Right; ++x)
                        ++histogram[block.Y, block.X, image[y, x]];
            }
            return histogram;
        }

        public short[, ,] SmoothAroundCorners(BlockMap blocks, short[, ,] input)
        {
            Point[] blocksAround = new Point[] { new Point(0, 0), new Point(-1, 0), new Point(0, -1), new Point(-1, -1) };
            short[, ,] output = new short[blocks.CornerCount.Height, blocks.CornerCount.Width, 256];
            foreach (var corner in blocks.AllCorners)
            {
                foreach (Point relative in blocksAround)
                {
                    Point block = Calc.Add(corner, relative);
                    if (blocks.AllBlocks.Contains(block))
                    {
                        for (int i = 0; i < 256; ++i)
                            output[corner.Y, corner.X, i] += input[block.Y, block.X, i];
                    }
                }
            }
            return output;
        }

        public short[, ,] Smooth(BlockMap blocks, short[, ,] input)
        {
            short[, ,] output = new short[blocks.CornerCount.Height, blocks.CornerCount.Width, 256];
            foreach (var corner in blocks.AllCorners)
            {
                for (int i = 0; i < 256; ++i)
                    output[corner.Y, corner.X, i] = input[corner.Y, corner.X, i];
                foreach (Point neigborRelative in Neighborhood.CornerNeighbors)
                {
                    Point neighbor = Calc.Add(corner, neigborRelative);
                    if (blocks.AllCorners.Contains(neighbor))
                    {
                        for (int i = 0; i < 256; ++i)
                            output[corner.Y, corner.X, i] += input[neighbor.Y, neighbor.X, i];
                    }
                }
            }
            return output;
        }
    }
}

File: java/sourceafis/src/sourceafis/extraction/filters/LocalHistogram.java
/**
 * @author Veaceslav Dubenco
 * @since 17.10.2012
 */
package sourceafis.extraction.filters;

import sourceafis.general.BlockMap;
import sourceafis.general.Calc;
import sourceafis.general.Neighborhood;
import sourceafis.general.Point;
import sourceafis.general.RectangleC;
import sourceafis.general.parallel.Parallel;
import sourceafis.general.parallel.ParallelForEachDelegate;

/**
 * 
 */
public final class LocalHistogram {

	public short[][][] Analyze(final BlockMap blocks, final byte[][] image) {
		final short[][][] histogram = new short[blocks.getBlockCount().Height][blocks
				.getBlockCount().Width][256];
		/*ParallelForEachDelegate<Point> delegate = new ParallelForEachDelegate<Point>() {
			@Override
			public Point delegate(Point block) {
				RectangleC area = blocks.getBlockAreas().get(block);
				for (int y = area.getBottom(); y < area.getTop(); ++y)
					for (int x = area.getLeft(); x < area.getRight(); ++x)
						++histogram[block.Y][block.X][image[y][x] & 0xFF];
				return null;
			}

			@Override
			public Point combineResults(Point res1, Point res2) {
				return null;
			}
		};
		Parallel.ForEach(blocks.getAllBlocks(), delegate);
		*/
		for (Point block : blocks.getAllBlocks()) {
			RectangleC area = blocks.getBlockAreas().get(block);
			for (int y = area.getBottom(); y < area.getTop(); ++y)
				for (int x = area.getLeft(); x < area.getRight(); ++x)
					++histogram[block.Y][block.X][image[y][x] & 0xFF];
		}
		return histogram;
	}

	public short[][][] SmoothAroundCorners(final BlockMap blocks,
			final short[][][] input) {
		final Point[] blocksAround = new Point[] { new Point(0, 0),
				new Point(-1, 0), new Point(0, -1), new Point(-1, -1) };
		final short[][][] output = new short[blocks.getCornerCount().Height][blocks
				.getCornerCount().Width][256];

		ParallelForEachDelegate<Point> delegate = new ParallelForEachDelegate<Point>() {
			@Override
			public Point delegate(Point corner) {
				for (Point relative : blocksAround) {
					Point block = Calc.Add(corner, relative);
					if (blocks.getAllBlocks().Contains(block)) {
						for (int i = 0; i < 256; ++i)
							output[corner.Y][corner.X][i] += input[block.Y][block.X][i];
					}
				}
				return null;
			}

			@Override
			public Point combineResults(Point res1, Point res2) {
				return null;
			}
		};

		Parallel.ForEach(blocks.getAllCorners(), delegate);
		return output;
	}

	public short[][][] Smooth(final BlockMap blocks, final short[][][] input) {
		final short[][][] output = new short[blocks.getCornerCount().Height][blocks
				.getCornerCount().Width][256];

		ParallelForEachDelegate<Point> delegate = new ParallelForEachDelegate<Point>() {
			@Override
			public Point delegate(Point corner) {
				for (int i = 0; i < 256; ++i)
					output[corner.Y][corner.X][i] = input[corner.Y][corner.X][i];
				for (Point neigborRelative : Neighborhood.CornerNeighbors) {
					Point neighbor = Calc.Add(corner, neigborRelative);
					if (blocks.getAllCorners().contains(neighbor)) {
						for (int i = 0; i < 256; ++i)
							output[corner.Y][corner.X][i] += input[neighbor.Y][neighbor.X][i];
					}
				}
				return null;
			}

			@Override
			public Point combineResults(Point res1, Point res2) {
				return null;
			}
		};

		Parallel.ForEach(blocks.getAllCorners(), delegate);
		return output;
	}
}
